File: src/fsharp/examples/nbody/NBody.OpenGL.fs
(**
Visualizing simulation using [OpenGL](https://www.opengl.org/) respectively [OpenTK](http://www.opentk.com/).
*)
(*** define:StartOpenGL ***)
module Tutorial.Fs.examples.NBody.OpenGL
open System
open System.Collections.Generic
open System.Diagnostics
open OpenTK
open OpenTK.Graphics
open OpenTK.Graphics.OpenGL
open Alea.CUDA

#nowarn "9"
#nowarn "51"

(**
We inhere from the [OpenTK](http://www.opentk.com/) class `GameWindow`.
*)
(*** define:startSimWindow ***)
type SimWindow() as this =
    inherit GameWindow(800, 600, GraphicsMode.Default, "Gravitational n-body simulation")

    let numBodies = 256*64
    let clusterScale = 1.0f
    let velocityScale = 1.0f
    //let initializeBodies = initializeBodies1 clusterScale velocityScale
    //let initializeBodies = initializeBodies2 clusterScale velocityScale
    let initializeBodies = initializeBodies3 clusterScale velocityScale
    let deltaTime = 0.001f
    let softeningSquared = 0.00125f
    let damping = 0.9995f
    let stopWatch = Stopwatch.StartNew()
    let fpsCalcLag = 128
    let mutable frameCounter = 0

(**
function to create cuda GL context, needed in order to initialize worker.
*)
(*** define:createGLContextGenerator ***)
    let generate() =
        let mutable cuContext = 0n
        let cuDevice = Device.Default
        cuSafeCall(cuGLCtxCreate(&&cuContext, 0u, Device.Default.ID))
        cuDevice, cuContext
(**
create worker using Cuda GL context generation function.
*)
(*** define:CreateWorker ***)
    // Note, we don't need worker.Eval cause we will run it in single thread.
    let worker =
        Worker.CreateOnCurrentThread(generate)

(**
Create several simulators: dynamic & static with different `blockSize`s for comparison.
*)
(***  define:CreateSimulatros ***)
    let simulators, disposeSimulators =
        let simulators = Queue<ISimulator>()
        let target = GPUModuleTarget.Worker(worker)

        let simulatorCPUSimpleModule = Impl.CPU.Simple.SimulatorModule()
        let simulatorGPUDynamicBlockSizeModule = new Impl.GPU.DynamicBlockSize.SimulatorModule(target)
        let simulatorGPUStaticBlockSizeModule64 = new Impl.GPU.StaticBlockSize.SimulatorModule64(target)
        let simulatorGPUStaticBlockSizeModule128 = new Impl.GPU.StaticBlockSize.SimulatorModule128(target)
        let simulatorGPUStaticBlockSizeModule256 = new Impl.GPU.StaticBlockSize.SimulatorModule256(target)
        let simulatorGPUStaticBlockSizeModule512 = new Impl.GPU.StaticBlockSize.SimulatorModule512(target)

        // first, enquene one simulator which is 256 blocksize so we can compare with C code for performance
        simulators.Enqueue(simulatorGPUStaticBlockSizeModule256.CreateSimulator())

        // now, enqueue several dynamic block size simulators
        simulators.Enqueue(simulatorGPUDynamicBlockSizeModule.CreateSimulator(64))
        simulators.Enqueue(simulatorGPUDynamicBlockSizeModule.CreateSimulator(128))
        simulators.Enqueue(simulatorGPUDynamicBlockSizeModule.CreateSimulator(256))
        simulators.Enqueue(simulatorGPUDynamicBlockSizeModule.CreateSimulator(512))

        // now, enqueue serveral static block size simulators
        simulators.Enqueue(simulatorGPUStaticBlockSizeModule64.CreateSimulator())
        simulators.Enqueue(simulatorGPUStaticBlockSizeModule128.CreateSimulator())
        simulators.Enqueue(simulatorGPUStaticBlockSizeModule256.CreateSimulator())
        simulators.Enqueue(simulatorGPUStaticBlockSizeModule512.CreateSimulator())

        // last, enqueue cpu simulator, this is quite slow, FPS = 0 :)
        //simulators.Enqueue(simulatorCPUSimpleModule.CreateSimulator(worker, numBodies))

        let dispose() =
            simulatorGPUDynamicBlockSizeModule.Dispose()
            simulatorGPUStaticBlockSizeModule64.Dispose()
            simulatorGPUStaticBlockSizeModule128.Dispose()
            simulatorGPUStaticBlockSizeModule256.Dispose()
            simulatorGPUStaticBlockSizeModule512.Dispose()

        simulators, dispose

    let mutable simulator = simulators.Dequeue()

(**
Method to describe simulation in window-title.
*)
(*** define:GLdescription***)
    let description() =
        let time = stopWatch.ElapsedMilliseconds
        let fps = (float frameCounter) * 1000.0 / (float time)
        this.Title <- sprintf "bodies %d, %s %A %d cores, %s, fps %f" numBodies worker.Device.Name worker.Device.Arch worker.Device.Cores simulator.Description fps
        stopWatch.Restart()

(**
Switch between different simulators in order to compare the performence for dynamic & static GPU implementation and different `blockSize`s.
*)
(*** define:GLswitchSimulators***)
    let switchSimulator() =
        simulators.Enqueue(simulator)
        simulator <- simulators.Dequeue()
        description()
        frameCounter <- 0
        stopWatch.Restart()

(**
Display options: exit simulation or switch simulator. 
*)
(*** define:GLhelp***)
    let help() =
        printfn "Press these keys:"
        printfn "[ESC]    : Exit"
        printfn "S        : Switch to next simulator"

(**
Create `buffer`s and `vel`, where the particles position and velocities will be stored in.
Positions are read from one `buffer` and stored to the other, hence the `buffer`s have to change between every integration step. This is what `swapPos` is for.
*)
(*** define:CreateBuffers***)
    let buffers = 
        let buffers = Array.zeroCreate<GLuint> 2
        GL.GenBuffers(buffers.Length, buffers)

        for buffer in buffers do
            GL.BindBuffer(BufferTarget.ArrayBuffer, buffer)            
            GL.BufferData(BufferTarget.ArrayBuffer, nativeint (sizeof<float4>*numBodies), (null : _ []), BufferUsageHint.DynamicDraw )
            let size = ref 0
            GL.GetBufferParameter(BufferTarget.ArrayBuffer, BufferParameterName.BufferSize, size)
            if !size <> sizeof<float4>*numBodies then
                failwith "Pixel Buffer Object allocation failed!"
            GL.BindBuffer(BufferTarget.ArrayBuffer, 0)
            cuSafeCall(cuGLRegisterBufferObject(buffer))

        buffers

    let resources = buffers |> Array.map (fun buffer ->
        let mutable res = 0n
        cuSafeCall(cuGraphicsGLRegisterBuffer(&&res, buffer, 0u))
        res)

    let mutable vel = worker.Malloc<float4>(numBodies)

    let swapPos() =
        let buffer' = buffers.[0]
        buffers.[0] <- buffers.[1]
        buffers.[1] <- buffer'

        let resource' = resources.[0]
        resources.[0] <- resources.[1]
        resources.[1] <- resource'

(**
Locks pointers from acces by [OpenTK](http://www.opentk.com/), s.t. function `f` can work on these pointers.
*)
(*** define:LockPositions ***)
    let lockPos(f:deviceptr<float4> -> deviceptr<float4> -> 'T) =
        cuSafeCall(cuGraphicsResourceSetMapFlags(resources.[0], uint32 CUgraphicsMapResourceFlags.CU_GRAPHICS_MAP_RESOURCE_FLAGS_READ_ONLY))
        cuSafeCall(cuGraphicsResourceSetMapFlags(resources.[1], uint32 CUgraphicsMapResourceFlags.CU_GRAPHICS_MAP_RESOURCE_FLAGS_WRITE_DISCARD))
        cuSafeCall(cuGraphicsMapResourcesEx(2u, resources, 0n))
        let mutable bytes = 0n
        let mutable handle0 = 0n
        let mutable handle1 = 0n
        cuSafeCall(cuGraphicsResourceGetMappedPointer(&&handle0, &&bytes, resources.[0]))
        cuSafeCall(cuGraphicsResourceGetMappedPointer(&&handle1, &&bytes, resources.[1]))
        let pos0 = deviceptr<float4>(handle0)
        let pos1 = deviceptr<float4>(handle1)
        try f pos0 pos1
        finally cuSafeCall(cuGraphicsUnmapResourcesEx(2u, resources, 0n))

(**
Final steps in order to initialize the simulaiton and the visualization.
*)
(*** define:FinalizeGL ***)
    do
        let hpos, hvel = initializeBodies numBodies
        worker.Scatter(hvel, vel.Ptr)
        lockPos <| fun pos0 pos1 -> worker.Scatter(hpos, pos1)
        help()
        description()

        this.KeyDown.Add(fun e ->
            match e.Key.ToString() with
            | "Escape" -> this.Exit()
            | "S" -> switchSimulator()
            | _ -> () )

(**
Override needed functionality from `GameWindow`.
*)
(*** define:overrideFuctions ***)
    override this.Dispose(disposing) =
        for resource in resources do cuSafeCall(cuGraphicsUnregisterResource(resource))
        for buffer in buffers do cuSafeCall(cuGLUnregisterBufferObject(buffer))
        if buffers.Length > 0 then GL.DeleteBuffers(buffers.Length, buffers)
        if disposing then vel.Dispose()
        if disposing then disposeSimulators()
        if disposing then worker.Dispose()
        base.Dispose(disposing)

    override this.OnLoad e =
        base.OnLoad(e)
        this.VSync <- VSyncMode.Off
        GL.ClearColor(0.0f, 0.0f, 0.0f, 0.0f) // black as the universe
        GL.Enable(EnableCap.DepthTest)

    override this.OnResize e =
        base.OnResize(e)
        GL.Viewport(this.ClientRectangle.X, this.ClientRectangle.Y, this.ClientRectangle.Width, this.ClientRectangle.Height)
        let projection = Matrix4.CreatePerspectiveFieldOfView((float32)Math.PI / 4.0f, float32 this.Width / (float32)this.Height, 1.0f, 64.0f)
        GL.MatrixMode(MatrixMode.Projection)
        GL.LoadMatrix(ref projection)

(**
Render function:

- Updates frame-per-second calculation and if needed description in title.
- Calls `Integrate` method from choosen `simulator`.
- Displays all the particles using [OpenTK](http://www.opentk.com/) functionality.
*)
(*** define:render ***)
    override this.OnRenderFrame e =
        base.OnRenderFrame(e)

        frameCounter <- frameCounter + 1
        if frameCounter >= fpsCalcLag then
            description()
            frameCounter <- 0

        swapPos()
        lockPos <| fun pos0 pos1 -> simulator.Integrate pos1 pos0 vel.Ptr numBodies deltaTime softeningSquared damping

        GL.Clear(ClearBufferMask.ColorBufferBit ||| ClearBufferMask.DepthBufferBit)
        let modelview = Matrix4.LookAt(Vector3.Zero, Vector3.UnitZ, Vector3.UnitY)
        GL.MatrixMode(MatrixMode.Modelview)
        GL.LoadMatrix(ref modelview)

        GL.Color3(1.0f, 215.0f/255.0f, 0.0f) // golden as the stars
        GL.EnableClientState(ArrayCap.VertexArray)
        GL.BindBuffer(BufferTarget.ArrayBuffer, buffers.[1])
        GL.VertexPointer(4, VertexPointerType.Float, 0, 0)
        GL.DrawArrays(PrimitiveType.Points, 0, numBodies)
        GL.DisableClientState(ArrayCap.VertexArray)

        GL.Finish()
        this.SwapBuffers()

(**
Defaults to use `SimWindow`.
*)
(*** define:runSimWindow ***)
let runSim() =
    use window = new SimWindow()
    window.Run()

let runPerformance() =
    Impl.GPU.DynamicBlockSize.Performance()
    Impl.GPU.StaticBlockSize.Performance()

File: src/csharp/examples/nbody/NBody.OpenGL.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Alea.CUDA;
using OpenTK;
using OpenTK.Graphics;
using OpenTK.Graphics.OpenGL;
using OpenTK.Input;

namespace Tutorial.Cs.examples.nbody
{
    //[startSimWindow]
    public delegate void Disposer();
    public delegate void Del(deviceptr<float4> a, deviceptr<float4> b);

    public class SimWindow : GameWindow
    {
        private readonly int _numBodies;
        private readonly float _deltaTime;
        private readonly float _softeningSquared;
        private readonly float _damping;
        private readonly Worker _worker;
        private readonly Queue<ISimulator> _simulators;
        private readonly Disposer _disposeSimulators;
        private ISimulator _simulator;
        private readonly uint[] _buffers;
        private readonly DeviceMemory<float4> _vel;
        private readonly IntPtr[] _resources;
        private readonly Stopwatch _stopwatch;
        private readonly int _fpsCalcLag;
        private int _frameCounter;
        //[/startSimWindow]

        //[createGLContextGenerator]
        unsafe Tuple<Device, IntPtr> Generate()
        {
            var cuContext = IntPtr.Zero;
            var cuDevice = Device.Default;
            CUDAInterop.cuSafeCall(CUDAInterop.cuGLCtxCreate(&cuContext, 0u, Device.Default.ID));
            return (new Tuple<Device, IntPtr>(cuDevice, cuContext));
        }
        //[/createGLContextGenerator]

        //[GLhelp]
        static void Help()
        {
            Console.WriteLine("Press these keys:");
            Console.WriteLine("[ESC]    : Exit");
            Console.WriteLine("S        : Switch to next simulator");
        }
        //[/GLhelp]

        //[GLdescription]
        void Description()
        {
            var time = _stopwatch.ElapsedMilliseconds;
            var fps = ((float) _frameCounter)*1000.0/((float) time);
            Title = string.Format("bodies {0}, {1} {2} {3} cores, {4}, fps {5}", _numBodies, _worker.Device.Name,
                _worker.Device.Arch, _worker.Device.Cores, _simulator.Description, fps);
            _stopwatch.Restart();
        }
        //[/GLdescription]

        //[LockPositions]
        void LockPos(Del f)
        {
            CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsResourceSetMapFlags(_resources[0],
                (uint) CUDAInterop.CUgraphicsMapResourceFlags_enum.CU_GRAPHICS_MAP_RESOURCE_FLAGS_READ_ONLY));
            CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsResourceSetMapFlags(_resources[1],
                (uint) CUDAInterop.CUgraphicsMapResourceFlags_enum.CU_GRAPHICS_MAP_RESOURCE_FLAGS_WRITE_DISCARD));
            CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsMapResourcesEx(2u, _resources, IntPtr.Zero));

            var bytes = IntPtr.Zero;
            var handle0 = IntPtr.Zero;
            var handle1 = IntPtr.Zero;
            unsafe
            {
                CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsResourceGetMappedPointer(&handle0, &bytes, 
                                                                                      _resources[0]));
                CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsResourceGetMappedPointer(&handle1, &bytes, 
                                                                                      _resources[1]));
            }
            var pos0 = new deviceptr<float4>(handle0);
            var pos1 = new deviceptr<float4>(handle1);
            try
            {
                f(pos0, pos1);
            }
            finally
            {
                CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsUnmapResourcesEx(2u, _resources, IntPtr.Zero));
            }
        }
        //[/LockPositions]

        public SimWindow() : base(800, 600, GraphicsMode.Default, "Gravitational n-body simulation")
        {
            _numBodies = 256*64;
            const float clusterScale = 1.0f;
            const float velocityScale = 1.0f;
            _deltaTime = 0.001f;
            _softeningSquared = 0.00125f;
            _damping = 0.9995f;
            //[CreateWorker]
            _worker = Worker.CreateOnCurrentThreadByFunc(Generate);
            //[/CreateWorker]

            _stopwatch = Stopwatch.StartNew();
            _fpsCalcLag = 128;
            _frameCounter = 0;

            //[CreateSimulatros]
            _simulators = new Queue<ISimulator>();
            var target = GPUModuleTarget.Worker(_worker);

            var simulatorGpuDynamicBlockSizeModule = new GpuDynamicSimulatorModule(target);         // need dispose
            var simulatorGpuDynamicBlockSize64 = simulatorGpuDynamicBlockSizeModule.Create(64);
            var simulatorGpuDynamicBlockSize128 = simulatorGpuDynamicBlockSizeModule.Create(128);
            var simulatorGpuDynamicBlockSize256 = simulatorGpuDynamicBlockSizeModule.Create(256);
            var simulatorGpuDynamicBlockSize512 = simulatorGpuDynamicBlockSizeModule.Create(512);

            var simulatorGpuStaticBlockSizeModule64 = new GpuStaticSimulatorModule64(target);       // need dispose
            var simulatorGpuStaticBlockSizeModule128 = new GpuStaticSimulatorModule128(target);     // need dispose
            var simulatorGpuStaticBlockSizeModule256 = new GpuStaticSimulatorModule256(target);     // need dispose
            var simulatorGpuStaticBlockSizeModule512 = new GpuStaticSimulatorModule512(target);     // need dispose

            // First, enquene one simulator which is 256 blocksize so we can compare with C code for performance.
            _simulators.Enqueue(simulatorGpuStaticBlockSizeModule256);

            // Enqueue several dynamic block size simulators.
            _simulators.Enqueue(simulatorGpuDynamicBlockSize64);
            _simulators.Enqueue(simulatorGpuDynamicBlockSize128);
            _simulators.Enqueue(simulatorGpuDynamicBlockSize256);
            _simulators.Enqueue(simulatorGpuDynamicBlockSize512);

            // Enqueue several static block size simulators.
            _simulators.Enqueue(simulatorGpuStaticBlockSizeModule64);
            _simulators.Enqueue(simulatorGpuStaticBlockSizeModule128);
            _simulators.Enqueue(simulatorGpuStaticBlockSizeModule256);
            _simulators.Enqueue(simulatorGpuStaticBlockSizeModule512);

            // We do not enqueue any cpu simulator as it is much too slow.
            //_simulators.Enqueue(new CpuSimulator(_worker, _numBodies));

            _disposeSimulators = () =>
            {
                simulatorGpuDynamicBlockSizeModule.Dispose();
                simulatorGpuStaticBlockSizeModule64.Dispose();
                simulatorGpuStaticBlockSizeModule128.Dispose();
                simulatorGpuStaticBlockSizeModule256.Dispose();
                simulatorGpuStaticBlockSizeModule512.Dispose();
            };

            _simulator = _simulators.Dequeue();
            //[/CreateSimulatros]

            //[CreateBuffers]
            _buffers = new uint[2];
            for (var i = 0; i < _buffers.Length; i++)
            {
                _buffers[i] = 0;
            }
            GL.GenBuffers(_buffers.Length, _buffers);
            foreach (var buffer in _buffers)
            {
                GL.BindBuffer(BufferTarget.ArrayBuffer, buffer);
                GL.BufferData(BufferTarget.ArrayBuffer,
                              (IntPtr) (Microsoft.FSharp.Core.Operators.SizeOf<float4>()*_numBodies), 
                              IntPtr.Zero, BufferUsageHint.DynamicDraw);
                var size = 0;
                unsafe
                {
                    GL.GetBufferParameter(BufferTarget.ArrayBuffer, BufferParameterName.BufferSize, &size);
                }
                if (size != Microsoft.FSharp.Core.Operators.SizeOf<float4>()*_numBodies)
                {
                    throw new Exception("Pixel Buffer Object allocation failed!");
                }
                GL.BindBuffer(BufferTarget.ArrayBuffer, 0);
                CUDAInterop.cuSafeCall(CUDAInterop.cuGLRegisterBufferObject(buffer));
            }

            _resources = new IntPtr[_buffers.Length];
            for (var i = 0; i < _buffers.Length; i++)
            {
                var res = IntPtr.Zero;
                unsafe
                {
                    CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsGLRegisterBuffer(&res, _buffers[i], 0u));
                }
                _resources[i] = res;
            }
            //[/CreateBuffers]

            //[FinalizeGL]
            _vel = _worker.Malloc<float4>(_numBodies);

            float4[] hpos, hvel;
            BodyInitializer.Initialize(new BodyInitializer3(), clusterScale, velocityScale, _numBodies, 
                                       out hpos, out hvel);
            _worker.Scatter(hvel, _vel.Ptr, Microsoft.FSharp.Core.FSharpOption<int>.None,
                Microsoft.FSharp.Core.FSharpOption<int>.None);
            LockPos(
                (pos0, pos1) =>
                    _worker.Scatter(hpos, pos1, Microsoft.FSharp.Core.FSharpOption<int>.None,
                        Microsoft.FSharp.Core.FSharpOption<int>.None));

            Help();
            Description();
            //[/FinalizeGL]
        }


        //[GLswitchSimulators]
        public void SwitchSimulator()
        {
            _simulators.Enqueue(_simulator);
            _simulator = _simulators.Dequeue();
            Description();
            _frameCounter = 0;
            _stopwatch.Restart();
        }
        //[/GLswitchSimulators]

        public void SwapPos()
        {
            var buffer = _buffers[0];
            _buffers[0] = _buffers[1];
            _buffers[1] = buffer;

            var resource = _resources[0];
            _resources[0] = _resources[1];
            _resources[1] = resource;
        }

        void HandleKeyDown(object sender, KeyboardKeyEventArgs e)
        {
            switch (e.Key)
            {
                case Key.Escape:
                    Exit();
                    break;
                case Key.S:
                    SwitchSimulator();
                    break;
            }
        }

        //[overrideFuctions]
        protected override void Dispose(bool disposing)
        {
            foreach (var resource in _resources)
            {
                CUDAInterop.cuSafeCall(CUDAInterop.cuGraphicsUnregisterResource(resource));
            }
            foreach (var buffer in _buffers)
            {
                CUDAInterop.cuSafeCall(CUDAInterop.cuGLUnregisterBufferObject(buffer));
            }
            if (_buffers.Length > 0)
            {
                GL.DeleteBuffers(_buffers.Length, _buffers);
            }
            if (disposing)
            {
                _vel.Dispose();
                _disposeSimulators();
                _worker.Dispose();
            }
            base.Dispose(disposing);
        }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            Keyboard.KeyDown += HandleKeyDown;
            VSync = VSyncMode.Off;
            GL.ClearColor(0.0f, 0.0f, 0.0f, 0.0f); // black as the universe
            GL.Enable(EnableCap.DepthTest);
        }
        
        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            GL.Viewport(ClientRectangle.X, ClientRectangle.Y, ClientRectangle.Width, ClientRectangle.Height);
            var projection = Matrix4.CreatePerspectiveFieldOfView((float) Math.PI/4.0f, 
                                                                  (float) Width/Height, 1.0f, 64.0f);
            GL.MatrixMode(MatrixMode.Projection);
            GL.LoadMatrix(ref projection);
        }
        //[/overrideFuctions]

        //[render]
        protected override void OnRenderFrame(FrameEventArgs e)
        {
            base.OnRenderFrame(e);

            _frameCounter++;
            if (_frameCounter >= _fpsCalcLag)
            {
                Description();
                _frameCounter = 0;
            }

            SwapPos();
            LockPos( (pos0,pos1) => _simulator.Integrate(pos1, pos0, _vel.Ptr, _numBodies, _deltaTime, 
                                                         _softeningSquared, _damping));

            GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
            var modelview = Matrix4.LookAt(Vector3.Zero, Vector3.UnitZ, Vector3.UnitY);
            GL.MatrixMode(MatrixMode.Modelview);
            GL.LoadMatrix(ref modelview);

            GL.Color3(1.0f, 215.0f/255.0f, 0.0f); // golden as the stars
            GL.EnableClientState(ArrayCap.VertexArray);
            GL.BindBuffer(BufferTarget.ArrayBuffer, _buffers[1]);
            GL.VertexPointer(4, VertexPointerType.Float, 0, 0);
            GL.DrawArrays(PrimitiveType.Points, 0, _numBodies);
            GL.DisableClientState(ArrayCap.VertexArray);

            GL.Finish();
            SwapBuffers();
        }
        //[/render]
    }

    //[runSimWindow]
    public static class Run
    {
        public static void Sim()
        {
            using (var window = new SimWindow())
            {
                window.Run();
            }
        }

        public static void Performance()
        {
            GpuDynamicSimulatorTests.Performance();
            GpuStaticSimulatorTests.Performence();
        }
    }
    //[/runSimWindow]
}